File: ai-service/app/services/order_service.py
from datetime import datetime
import uuid
from bson import ObjectId

from app.database import get_database

from app.repositories.order_repository import OrderRepository
from app.repositories.product_repository import ProductRepository
from app.repositories.customer_repository import CustomerRepository

from app.services.setting_service import SettingService
from app.services.inventory_history_service import InventoryHistoryService



class OrderService:


    def __init__(self):

        self.order_repository = OrderRepository()

        self.product_repository = ProductRepository()

        self.customer_repository = CustomerRepository()

        self.setting_service = SettingService()

        self.inventory_history = InventoryHistoryService()



    async def generate_order_id(self):

        return "ORD-" + uuid.uuid4().hex[:8].upper()




    async def create_order(

        self,

        customer_id,

        items,

        shipping_address,

        city,

        province,

        payment_method="Cash on Delivery",

        notes=""

    ):


        database = get_database()

        client = database.client


        session = await client.start_session()



        try:


            async with session.start_transaction():


                order_items = []

                subtotal = 0

                total_discount = 0



                # =========================
                # Validate Products
                # =========================


                for item in items:


                    product = await self.product_repository.get_by_id(
                        item["product"]
                    )


                    if not product:

                        raise Exception(
                            "Product not found"
                        )



                    quantity = item.get(
                        "quantity",
                        1
                    )



                    if product["stock"] < quantity:

                        raise Exception(
                            f"{product['productName']} is out of stock"
                        )



                    price = product["price"]


                    discount_percentage = product.get(
                        "discount",
                        0
                    )



                    discount_amount = (
                        price *
                        discount_percentage /
                        100
                    )



                    final_price = (
                        price -
                        discount_amount
                    )



                    item_total = (
                        final_price *
                        quantity
                    )



                    subtotal += item_total


                    total_discount += (
                        discount_amount *
                        quantity
                    )



                    order_items.append({

                        "product": product["_id"],

                        "productName":
                            product["productName"],

                        "quantity":
                            quantity,

                        "selectedSize":
                            item.get("selectedSize",""),

                        "selectedColor":
                            item.get("selectedColor",""),

                        "originalPrice":
                            price,

                        "discountPercentage":
                            discount_percentage,

                        "discountAmount":
                            discount_amount,

                        "finalPrice":
                            final_price,

                        "subtotal":
                            item_total

                    })



                # =========================
                # Delivery Calculation
                # =========================


                delivery_data = await self.setting_service.calculate_delivery_charge(

                    city,

                    province

                )


                delivery_charge = delivery_data["charge"]



                delivery_settings = await self.setting_service.get_delivery_settings()



                if subtotal >= delivery_settings["freeDeliveryAbove"]:

                    delivery_charge = 0



                grand_total = (

                    subtotal +

                    delivery_charge

                )



                # =========================
                # Create Order
                # =========================


                order = {


                    "orderId":
                        await self.generate_order_id(),


                    "customer":
                        customer_id,


                    "products":
                        order_items,


                    "subtotal":
                        subtotal,


                    "totalDiscount":
                        total_discount,


                    "deliveryCharges":
                        delivery_charge,


                    "grandTotal":
                        grand_total,


                    "paymentMethod":
                        payment_method,


                    "paymentStatus":
                        "Pending",


                    "status":
                        "Pending",


                    "trackingNumber":
                        "",


                    "shippingAddress":
                        shipping_address,


                    "city":
                        city,


                    "province":
                        province,


                    "notes":
                        notes,


                    "createdAt":
                        datetime.utcnow(),


                    "updatedAt":
                        datetime.utcnow()

                }



                saved_order = await self.order_repository.create(

                    order,

                    session=session

                )



                # =========================
                # Reduce Stock
                # =========================


                for item in items:


                    await self.product_repository.update_stock(

                        item["product"],

                        item["quantity"],

                        session=session

                    )



                # =========================
                # Customer History
                # =========================


                await self.customer_repository.add_order_history(

                    customer_id,

                    saved_order["_id"],

                    session=session

                )


                return saved_order




        except Exception as error:


            await session.abort_transaction()

            raise error



        finally:

            await session.end_session()




    async def get_order(self, order_id):

        return await self.order_repository.get_by_id(
            order_id
        )



    async def get_customer_orders(self, customer_id):

        return await self.order_repository.get_by_customer(
            customer_id
        )



    async def update_status(self, order_id, status):

        return await self.order_repository.update_status(
            order_id,
            status
        )



    async def update_tracking(self, order_id, tracking_number):

        return await self.order_repository.update_tracking(
            order_id,
            tracking_number
        )
        
        
    async def get_orders(self):

     return await self.order_repository.get_all()  
 
 
    async def update_order(
     self,
     order_id,
     data
    ):

     return await self.order_repository.update(
        order_id,
        data
    )  
     
     
     
    async def delete_order(
     self,
     order_id
    ):

     return await self.order_repository.delete(
        order_id
    )
     
     
     
    async def get_order_by_number(
     self,
     order_number
    ):

     return await self.order_repository.get_by_order_number(
        order_number
    ) 
     
     
    async def cancel_order(
    self,
    order_number
):

     return await self.order_repository.cancel_order(
        order_number
    )

    async def get_order_by_tracking(
        self,
        tracking_number
    ):
        return await self.order_repository.get_by_tracking(
            tracking_number
        )
    
    
    async def cancel_order_with_restore(
        self,
        order_number
    ):
        order = await self.order_repository.get_by_order_number(order_number)
        if not order:
            raise ValueError(f"Order {order_number} not found.")
        if order.get("status") == "Cancelled":
            raise ValueError(f"Order {order_number} is already cancelled.")

        database = get_database()
        client = database.client
        session = await client.start_session()

        try:
            async with session.start_transaction():
                for item in order.get("products", []):
                    product_id = str(item.get("product", ""))
                    if product_id and ObjectId.is_valid(product_id):
                        product = await self.product_repository.get_by_id(product_id)
                        if product:
                            qty_before = product["stock"]
                            await self.product_repository.increase_stock(
                                product_id, item.get("quantity", 1), session=session
                            )
                            await self.inventory_history.record_cancellation(
                                order_id=order["orderId"],
                                product_id=product_id,
                                product_name=item.get("productName", ""),
                                quantity_before=qty_before,
                                quantity_after=qty_before + item.get("quantity", 1),
                                quantity_restored=item.get("quantity", 1),
                            )

                await self.order_repository.update_status(
                    order_number, "Cancelled", session=session
                )

            return await self.order_repository.get_by_order_number(order_number)

        except Exception as error:
            await session.abort_transaction()
            raise error
        finally:
            await session.end_session()

File: backend/services/orderService.js
// const mongoose = require("mongoose");

// const Order = require("../models/Order");
// const Product = require("../models/Product");
// const Customer = require("../models/Customer");

// // ==============================
// // Create Order
// // ==============================

// const createOrderService = async (orderData) => {

//     const session = await mongoose.startSession();

//     session.startTransaction();

//     try {

//         // ==========================
//         // Generate Order Number
//         // ==========================

//         const lastOrder = await Order.findOne()
//             .sort({ createdAt: -1 })
//             .session(session);

//         let nextNumber = 1001;

//         if (lastOrder && lastOrder.orderId) {

//             const lastId = parseInt(lastOrder.orderId.split("-")[1]);

//             nextNumber = lastId + 1;

//         }

//         const orderItems = [];

//         let subtotal = 0;

//         let totalDiscount = 0;

//         // ==========================
//         // Process Every Product
//         // ==========================

//         for (const item of orderData.products) {

//             const product = await Product.findById(item.product).session(session);

//             if (!product) {

//                 throw new Error(`Product not found : ${item.product}`);

//             }

//             // Product Disabled

//             if (!product.status) {

//                 throw new Error(`${product.productName} is unavailable`);

//             }

//             // Stock Check

//             if (product.stock < item.quantity) {

//                 throw new Error(
//                     `${product.productName} has only ${product.stock} item(s) left`
//                 );

//             }

//             // ==========================
//             // Price Calculation
//             // ==========================

//             const originalPrice = product.price;

//             const discountPercentage = product.discount || 0;

//             const discountAmount =
//                 (originalPrice * discountPercentage) / 100;

//             const finalPrice =
//                 originalPrice - discountAmount;

//             const itemSubtotal =
//                 finalPrice * item.quantity;

//             subtotal += originalPrice * item.quantity;

//             totalDiscount += discountAmount * item.quantity;

//             // ==========================
//             // Reduce Stock
//             // ==========================

//             product.stock -= item.quantity;

//             await product.save({ session });

//             // ==========================
//             // Save Item
//             // ==========================

//             orderItems.push({

//                 product: product._id,

//                 productName: product.productName,

//                 quantity: item.quantity,

//                 selectedSize: item.selectedSize,

//                 selectedColor: item.selectedColor,

//                 originalPrice,

//                 discountPercentage,

//                 discountAmount,

//                 finalPrice,

//                 subtotal: itemSubtotal,

//             });

//         }

//                 // ==========================
//         // Delivery Charges
//         // ==========================

//         const deliveryCharges = orderData.deliveryCharges || 0;

//         // Customer pays after discount
//         const grandTotal =
//             subtotal - totalDiscount + deliveryCharges;

//         // ==========================
//         // Check Customer
//         // ==========================

//         const customer = await Customer.findById(
//             orderData.customer
//         ).session(session);

//         if (!customer) {

//             throw new Error("Customer not found");

//         }

//         // ==========================
//         // Create Order
//         // ==========================

//         const order = await Order.create(
//             [
//                 {

//                     orderId: `ORD-${nextNumber}`,

//                     customer: customer._id,

//                     products: orderItems,

//                     subtotal,

//                     totalDiscount,

//                     deliveryCharges,

//                     grandTotal,

//                     paymentMethod: orderData.paymentMethod,

//                     paymentStatus:
//                         orderData.paymentStatus || "Pending",

//                     status:
//                         orderData.status || "Pending",

//                     shippingAddress:
//                         orderData.shippingAddress,

//                     city: orderData.city,

//                     notes: orderData.notes || "",

//                 },
//             ],
//             { session }
//         );

//         // ==========================
//         // Add Order To Customer History
//         // ==========================

//         customer.orderHistory.push(order[0]._id);

//         await customer.save({ session });

//         // ==========================
//         // Commit Transaction
//         // ==========================

//         await session.commitTransaction();

//         session.endSession();

//         return order[0];

//     } catch (error) {

//         // ==========================
//         // Rollback
//         // ==========================

//         await session.abortTransaction();

//         session.endSession();

//         throw error;

//     }

// };

// module.exports = {
//     createOrderService,
// }; 



const mongoose = require("mongoose");

const Order = require("../models/Order");
const Product = require("../models/Product");
const Customer = require("../models/Customer");
const Setting = require("../models/Setting");


// ==============================
// Create Order Service
// ==============================

const createOrderService = async (orderData) => {


    const session = await mongoose.startSession();


    session.startTransaction();


    try {


        // ==============================
        // Generate Order ID
        // ==============================


        const lastOrder = await Order.findOne()
            .sort({ createdAt: -1 })
            .session(session);



        let nextNumber = 1001;



        if(lastOrder && lastOrder.orderId){


            const lastId =
                parseInt(
                    lastOrder.orderId.split("-")[1]
                );


            nextNumber = lastId + 1;

        }





        const orderItems = [];


        let subtotal = 0;


        let totalDiscount = 0;





        // ==============================
        // Process Products
        // ==============================


        for(const item of orderData.products){



            const product =
                await Product.findById(
                    item.product
                )
                .session(session);




            if(!product){


                throw new Error(
                    `Product not found : ${item.product}`
                );


            }




            // Product Status Check

            if(!product.status){


                throw new Error(
                    `${product.productName} is unavailable`
                );


            }





            // Stock Check

            if(product.stock < item.quantity){


                throw new Error(
                    `${product.productName} has only ${product.stock} item(s) left`
                );


            }





            // ==============================
            // Price Calculation
            // ==============================


            const originalPrice =
                product.price;



            const discountPercentage =
                product.discount || 0;




            const discountAmount =
                (originalPrice * discountPercentage) / 100;




            const finalPrice =
                originalPrice - discountAmount;




            const itemSubtotal =
                finalPrice * item.quantity;





            subtotal +=
                originalPrice * item.quantity;




            totalDiscount +=
                discountAmount * item.quantity;





            // ==============================
            // Reduce Stock
            // ==============================


            product.stock -= item.quantity;



            await product.save({
                session
            });





            // ==============================
            // Add Order Item
            // ==============================


            orderItems.push({


                product:
                    product._id,


                productName:
                    product.productName,


                quantity:
                    item.quantity,


                selectedSize:
                    item.selectedSize,


                selectedColor:
                    item.selectedColor,



                originalPrice,


                discountPercentage,


                discountAmount,


                finalPrice,


                subtotal:
                    itemSubtotal


            });



        }






        // ==============================
        // Delivery Charges
        // ==============================


        const settings =
            await Setting.findOne()
            .session(session);



        if(!settings){


            throw new Error(
                "Store settings not configured"
            );


        }



        let deliveryCharges = 0;



        const finalAmount =
            subtotal - totalDiscount;






        // Free Delivery


        if(
            settings.freeDeliveryAbove &&
            finalAmount >= settings.freeDeliveryAbove
        ){


            deliveryCharges = 0;


        }



        else {



            // Same City


            if(

                orderData.city &&
                orderData.city.toLowerCase()
                ===
                settings.storeCity.toLowerCase()

            ){


                deliveryCharges =
                    settings.sameCityCharge;


            }



            // Same Province


            else if(

                orderData.province &&
                orderData.province.toLowerCase()
                ===
                settings.storeProvince.toLowerCase()

            ){


                deliveryCharges =
                    settings.sameProvinceCharge;


            }



            // Other Province


            else{


                deliveryCharges =
                    settings.otherProvinceCharge;


            }



        }






        // Manual Override

        if(orderData.deliveryCharges !== undefined){


            deliveryCharges =
                Number(orderData.deliveryCharges);


        }




        if(deliveryCharges < 0){


            throw new Error(
                "Invalid delivery charges"
            );


        }







        // ==============================
        // Grand Total
        // ==============================


        const grandTotal =
            subtotal -
            totalDiscount +
            deliveryCharges;









        // ==============================
        // Check Customer
        // ==============================


        const customer =
            await Customer.findById(
                orderData.customer
            )
            .session(session);





        if(!customer){


            throw new Error(
                "Customer not found"
            );


        }









        // ==============================
        // Create Order
        // ==============================


        const order =
            await Order.create(
            [
                {


                    orderId:
                        `ORD-${nextNumber}`,



                    customer:
                        customer._id,



                    products:
                        orderItems,



                    subtotal,



                    totalDiscount,



                    deliveryCharges,



                    grandTotal,



                    paymentMethod:
                        orderData.paymentMethod,



                    paymentStatus:
                        orderData.paymentStatus ||
                        "Pending",



                    status:
                        orderData.status ||
                        "Pending",



                    shippingAddress:
                        orderData.shippingAddress,



                    city:
                        orderData.city,



                    province:
                        orderData.province,



                    notes:
                        orderData.notes || ""



                }

            ],
            {
                session
            });









        // ==============================
        // Update Customer History
        // ==============================


        customer.orderHistory.push(
            order[0]._id
        );



        await customer.save({
            session
        });









        // ==============================
        // Commit Transaction
        // ==============================


        await session.commitTransaction();



        session.endSession();




        return order[0];



    }



    catch(error){



        await session.abortTransaction();



        session.endSession();



        throw error;


    }



};






module.exports = {

    createOrderService

};